opentelemetry-sdk/src/resource/mod.rs

```diff
@@ -34,13 +34,22 @@ use opentelemetry::{Key, KeyValue, Value};
 use std::borrow::Cow;
 use std::collections::{hash_map, HashMap};
 use std::ops::Deref;
+use std::sync::Arc;
 use std::time::Duration;
 
+/// Inner structure of `Resource` holding the actual data.
```

Comment:
while this does improve the perf very nicely, I think we should further explore other ideas from https://github.com/open-telemetry/opentelemetry-rust/issues/1124

Even with this PR, we are still passing the Resource to each span, for every invocation. (for cases like simple/re-entrant processor, this is still unacceptable). We should explore is Processors/Exporters, who cares about Resource, can ask its provider for the Resource, if it wants, instead of passing it every time. 

This approach is used in OTel .NET : https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.Console/ConsoleActivityExporter.cs#L115

Not a blocker for this PR, but this is definitely something worth continuing to fine tune further.

Reply:
> Even with this PR, we are still passing the Resource to each span, for every invocation

Resource is only passed to processor after cloning, as per the code here - https://github.com/open-telemetry/opentelemetry-rust/blob/3f327a1ecbe9db3ad67323565c3347884ea5a2d3/opentelemetry-sdk/src/trace/span.rs#L206

Reply:
> Even with this PR, we are still **passing the Resource to each span**, for every invocation. (for cases like simple/re-entrant processor, this is still unacceptable). We should explore if Processors/Exporters, who cares about Resource, can ask its provider for the Resource, if it wants, instead of passing it every time. 
> 

Yea this seems like it would be a thing to remove. Especially since it's something that is configured per provider anyhow.

Reply:
> essor after cloning, as per the code here -

Yes. What I was saying was - there is no need to even pass the cloned one (however cheap cloning is) to the processors each time! The size of the struct passed to the processors can be reduced by chosing not to include Resource in it!.

Agree this is a follow up candidate.

Reply:
@TommyCpp also shared possibly only passing the Tracer, from which instrumentation-scope can be obtained, and from which Resource can also be obtained by first obtaining the TracerProvider. Will table this and come back later.